custom-metrics-stackdriver-adapter/pkg/adapter/translator/utils/filter_builder.go

```diff
@@ -0,0 +1,215 @@
+//
+// Example:
+//
+//	// To add "resource.type = \"custom.googleapis.com/foo\""
+//	filterBuilder := NewFilterBuilder(SchemaTypes[PodSchemaKey]).WithMetricType("custom.googleapis.com/foo")
+func (fb *FilterBuilder) WithMetricType(metricType string) *FilterBuilder {
+	fb.filters = append(fb.filters, fmt.Sprintf("%s = %q", fb.schema.metricType, metricType))
+	return fb
+}
+
+// WithProject adds a filter for project id.
+//
+// Example:
+//
+//	// To add "resource.labels.project_id = \"my-project\""
+//	filterBuilder := NewFilterBuilder(SchemaTypes[PodSchemaKey]).WithProject("my-project")
+func (fb *FilterBuilder) WithProject(project string) *FilterBuilder {
+	fb.filters = append(fb.filters, fmt.Sprintf("%s = %q", fb.schema.project, project))
+	return fb
+}
+
+// WithCluster adds a filter for cluster name.
+//
+// Example:
+//
+//	// To add "resource.labels.cluster_name = \"my-cluster\""
+//	filterBuilder := NewFilterBuilder(SchemaTypes[PodSchemaKey]).WithCluster("my-cluster")
+func (fb *FilterBuilder) WithCluster(cluster string) *FilterBuilder {
+	fb.filters = append(fb.filters, fmt.Sprintf("%s = %q", fb.schema.cluster, cluster))
+	return fb
+}
+
+// WithLocation adds a filter for location.
+//
+// Example:
+//
+//	// To add "resource.labels.location = \"my-location\""
+//	filterBuilder := NewFilterBuilder(SchemaTypes[PodSchemaKey]).WithLocation("my-location")
+func (fb *FilterBuilder) WithLocation(location string) *FilterBuilder {
+	fb.filters = append(fb.filters, fmt.Sprintf("%s = %q", fb.schema.location, location))
+	return fb
+}
+
+// WithContainer adds a filter for container. (used ONLY in the legacy model)
+//
+// Example:
+//
+//	// To add "resource.labels.container_name = \"my-container\""
+//	filterBuilder := NewFilterBuilder(SchemaTypes[PodSchemaKey]).WithContainer("my-container")
+func (fb *FilterBuilder) WithContainer() *FilterBuilder {
+	fb.filters = append(fb.filters, fmt.Sprintf("resource.labels.container_name = %q", ""))
+	return fb
+}
+
+// WithNamespace adds a filter for namespace.
+// (note: empty namespace will be ignored)
+//
+// Example:
+//
+//	// To add "resource.labels.namespace_name = \"my-namepace\""
+//	filterBuilder := NewFilterBuilder(SchemaTypes[PodSchemaKey]).WithNamespace("my-namespace")
+func (fb *FilterBuilder) WithNamespace(namespace string) *FilterBuilder {
+	if namespace != "" {
+		fb.filters = append(fb.filters, fmt.Sprintf("%s = %q", fb.schema.namespace, namespace))
+	}
+	return fb
+}
+
+// WithPods adds a filter for pods.
+//
+// Example:
+//
+//  1. To add "resource.labels.pod_name = my-pod"
+//     filterBuilder := NewFilterBuilder(SchemaTypes[PodSchemaKey]).WithPods([]string{"my-pod"})
+//
+//  2. To add "resource.lables.pod_name = one_of(my-pod-1,my-pod-2)"
+//     filterBuilder := NewFilterBuilder(SchemaTypes[PodSchemaKey]).WithPods([]string{"my-pod-1", "my-pod-2"})
+func (fb *FilterBuilder) WithPods(pods []string) *FilterBuilder {
+	if len(pods) == 0 {
```

Comment:
It can be 

Option 1: Switch
```
switch len(pods) {
  case 0: 
      klog.Warningf("...")
  case 1:
      fb.filters = append(...)
  default:
      fb.filters = append(...)
}

return fb
```
Option 2: If and return
```
if len(pods) ==0 {
  klog....
  return fb
}
if len(pods) == 1 {
  fb.filters = ...
  return fbs
}
fb.filters = ...
return fbs
```